File: src/parser/evaluator.rs
use super::ast::{Clafer, Declaration, Element, Expr};
use anyhow::{Result, anyhow};
use std::collections::{HashMap, HashSet};

/// A resolved instance: each leaf feature classified as included or excluded.
/// Structural containers and the abstract supertype appear in neither set.
#[derive(Debug, Default, PartialEq, Eq)]
pub struct Configuration {
    pub included: HashSet<String>,
    pub excluded: HashSet<String>,
}

/// Per-clafer state during resolution: active / inactive / undetermined.
type Assignment = HashMap<String, Option<bool>>;

pub struct Evaluator {
    // Global registry of every Clafer definition, keyed by name.
    pub registry: HashMap<String, Clafer>,
}

impl Evaluator {
    pub fn new(decls: Vec<Declaration>) -> Self {
        let mut registry = HashMap::new();

        // Index nested clafers too (e.g. SQLite inside Storage).
        fn index_recursive(elements: &[Element], map: &mut HashMap<String, Clafer>) {
            for el in elements {
                if let Element::Clafer(c) = el {
                    map.insert(c.name.clone(), c.clone());
                    index_recursive(&c.children, map);
                }
            }
        }

        for decl in decls {
            if let Declaration::Element(Element::Clafer(c)) = decl {
                registry.insert(c.name.clone(), c.clone());
                index_recursive(&c.children, &mut registry);
            }
        }

        Self { registry }
    }

    /// Resolves an instance against its abstract model into included/excluded leaf features.
    pub fn resolve_instance(&self, instance_name: &str) -> Result<Configuration> {
        let instance = self
            .registry
            .get(instance_name)
            .ok_or_else(|| anyhow!("Instance '{}' not found in model", instance_name))?;
        let super_name = instance
            .super_type
            .as_ref()
            .ok_or_else(|| anyhow!("Instance '{}' has no super type", instance_name))?;
        let supertype = self
            .registry
            .get(super_name)
            .ok_or_else(|| anyhow!("Super type '{}' of instance '{}' not found", super_name, instance_name))?;

        let mut subtree_names = HashSet::new();
        self.collect_subtree_names(supertype, &mut subtree_names);

        let mut assignment: Assignment = subtree_names.iter().map(|n| (n.clone(), None)).collect();
        try_assign(&mut assignment, super_name, true);

        // Seed explicit instance-body selections first so they win over inference.
        let mut instance_constraints = Vec::new();
        collect_constraints(&instance.children, &mut instance_constraints);
        for expr in &instance_constraints {
            propagate_expr(expr, true, &mut assignment);
        }

        // All subtree constraints for the fixed-point loop; re-asserting the
        // instance constraints is a harmless no-op.
        let mut constraints = instance_constraints;
        collect_constraints(&supertype.children, &mut constraints);

        loop {
            let mut changed = false;
            changed |= self.propagate_mandatory_children(&subtree_names, &mut assignment);
            changed |= self.propagate_group_cardinality(&subtree_names, &mut assignment);
            for expr in &constraints {
                changed |= propagate_expr(expr, true, &mut assignment);
            }
            if !changed {
                break;
            }
        }

        let mut config = Configuration::default();
        for name in &subtree_names {
            if name == super_name {
                continue;
            }
            let Some(clafer) = self.registry.get(name) else {
                continue;
            };
            if clafer.is_abstract {
                continue;
            }
            let is_structural_leaf = !clafer.children.iter().any(|e| matches!(e, Element::Clafer(_)));
            if !is_structural_leaf {
                continue;
            }
            match assignment.get(name).copied().flatten() {
                Some(true) => {
                    config.included.insert(name.clone());
                }
                _ => {
                    config.excluded.insert(name.clone());
                }
            }
        }
        Ok(config)
    }

    /// Collects every clafer name nested under `root` (inclusive), bounding the
    /// resolution so unrelated declarations elsewhere don't leak in.
    fn collect_subtree_names(&self, root: &Clafer, into: &mut HashSet<String>) {
        into.insert(root.name.clone());
        for el in &root.children {
            if let Element::Clafer(child) = el {
                self.collect_subtree_names(child, into);
            }
        }
    }

    /// For each active clafer without a `gcard`, forces its mandatory-`card`
    /// children active. Clafers with a `gcard` are handled by group-cardinality rules.
    fn propagate_mandatory_children(&self, subtree_names: &HashSet<String>, assignment: &mut Assignment) -> bool {
        let mut changed = false;
        for name in subtree_names {
            if assignment.get(name).copied().flatten() != Some(true) {
                continue;
            }
            let Some(clafer) = self.registry.get(name) else { continue };
            if clafer.gcard.is_some() {
                continue;
            }
            for el in &clafer.children {
                if let Element::Clafer(child) = el
                    && is_mandatory_card(&child.card)
                {
                    changed |= try_assign(assignment, &child.name, true);
                }
            }
        }
        changed
    }

    /// For each active clafer with a `gcard`, enforces its min/max bound: at max,
    /// remaining unknown children go false; when all are needed for min, they go
    /// true. Covers `xor`/`mux`/`or`/`opt`/explicit ncard.
    fn propagate_group_cardinality(&self, subtree_names: &HashSet<String>, assignment: &mut Assignment) -> bool {
        let mut changed = false;
        for name in subtree_names {
            if assignment.get(name).copied().flatten() != Some(true) {
                continue;
            }
            let Some(clafer) = self.registry.get(name) else { continue };
            let Some(gcard) = &clafer.gcard else { continue };
            let (min, max) = gcard_bounds(gcard);

            let direct_children: Vec<&str> = clafer
                .children
                .iter()
                .filter_map(|e| if let Element::Clafer(c) = e { Some(c.name.as_str()) } else { None })
                .collect();

            let true_count = direct_children
                .iter()
                .filter(|c| assignment.get(**c).copied().flatten() == Some(true))
                .count();
            let unknown: Vec<&str> = direct_children
                .iter()
                .filter(|c| assignment.get(**c).copied().flatten().is_none())
                .copied()
                .collect();

            if let Some(max) = max
                && true_count >= max
            {
                for child in &unknown {
                    changed |= try_assign(assignment, child, false);
                }
                continue;
            }
            if min > 0 && true_count + unknown.len() == min {
                for child in &unknown {
                    changed |= try_assign(assignment, child, true);
                }
            }
        }
        changed
    }
}

/// Gathers every `Element::Constraint` in `elements`, including nested clafers.
fn collect_constraints(elements: &[Element], out: &mut Vec<Expr>) {
    for el in elements {
        match el {
            Element::Constraint(expr) => out.push(expr.clone()),
            Element::Clafer(c) => collect_constraints(&c.children, out),
        }
    }
}

/// Sets `name` to `value` only if currently unknown; returns whether it changed.
/// Conflicts are dropped with a warning, keeping assignments monotonic
/// (`None -> Some` only) so the resolution loop terminates and explicit selections win.
fn try_assign(assignment: &mut Assignment, name: &str, value: bool) -> bool {
    match assignment.get(name).copied() {
        Some(Some(existing)) if existing == value => false,
        Some(Some(existing)) => {
            eprintln!(
                "warning: conflicting assignment for '{}': already {}, ignoring {}",
                name, existing, value
            );
            false
        }
        Some(None) => {
            assignment.insert(name.to_string(), Some(value));
            true
        }
        None => {
            eprintln!("warning: reference to unknown feature '{}'", name);
            false
        }
    }
}

/// Three-valued evaluation of an already-known (partial) assignment.
fn eval_expr(expr: &Expr, assignment: &Assignment) -> Option<bool> {
    match expr {
        Expr::Ref(name) => assignment.get(name).copied().flatten(),
        Expr::Not(a) => eval_expr(a, assignment).map(|v| !v),
        Expr::And(a, b) => match (eval_expr(a, assignment), eval_expr(b, assignment)) {
            (Some(false), _) | (_, Some(false)) => Some(false),
            (Some(true), Some(true)) => Some(true),
            _ => None,
        },
        Expr::Or(a, b) => match (eval_expr(a, assignment), eval_expr(b, assignment)) {
            (Some(true), _) | (_, Some(true)) => Some(true),
            (Some(false), Some(false)) => Some(false),
            _ => None,
        },
        Expr::Implies(l, r) => {
            let lv = eval_expr(l, assignment);
            let rv = eval_expr(r, assignment);
            if lv == Some(false) || rv == Some(true) {
                Some(true)
            } else if lv == Some(true) && rv == Some(false) {
                Some(false)
            } else {
                None
            }
        }
        Expr::Xor(a, b) => match (eval_expr(a, assignment), eval_expr(b, assignment)) {
            (Some(av), Some(bv)) => Some(av != bv),
            _ => None,
        },
        Expr::Iff(a, b) => match (eval_expr(a, assignment), eval_expr(b, assignment)) {
            (Some(av), Some(bv)) => Some(av == bv),
            _ => None,
        },
        Expr::Other(_) => None,
    }
}

/// Pushes a required truth value down through `expr`, asserting whatever can be
/// inferred about its subexpressions. Returns whether any assignment changed.
fn propagate_expr(expr: &Expr, required: bool, assignment: &mut Assignment) -> bool {
    match expr {
        Expr::Ref(name) => try_assign(assignment, name, required),
        Expr::Not(a) => propagate_expr(a, !required, assignment),
        Expr::And(a, b) => {
            if required {
                propagate_expr(a, true, assignment) | propagate_expr(b, true, assignment)
            } else {
                let mut changed = false;
                if eval_expr(a, assignment) == Some(true) {
                    changed |= propagate_expr(b, false, assignment);
                }
                if eval_expr(b, assignment) == Some(true) {
                    changed |= propagate_expr(a, false, assignment);
                }
                changed
            }
        }
        Expr::Or(a, b) => {
            if required {
                let mut changed = false;
                if eval_expr(a, assignment) == Some(false) {
                    changed |= propagate_expr(b, true, assignment);
                }
                if eval_expr(b, assignment) == Some(false) {
                    changed |= propagate_expr(a, true, assignment);
                }
                changed
            } else {
                propagate_expr(a, false, assignment) | propagate_expr(b, false, assignment)
            }
        }
        Expr::Implies(l, r) => {
            if required {
                let mut changed = false;
                if eval_expr(l, assignment) == Some(true) {
                    changed |= propagate_expr(r, true, assignment);
                }
                if eval_expr(r, assignment) == Some(false) {
                    changed |= propagate_expr(l, false, assignment);
                }
                changed
            } else {
                propagate_expr(l, true, assignment) | propagate_expr(r, false, assignment)
            }
        }
        Expr::Xor(a, b) => {
            let mut changed = false;
            if let Some(av) = eval_expr(a, assignment) {
                changed |= propagate_expr(b, av != required, assignment);
            }
            if let Some(bv) = eval_expr(b, assignment) {
                changed |= propagate_expr(a, bv != required, assignment);
            }
            changed
        }
        Expr::Iff(a, b) => {
            let mut changed = false;
            if let Some(av) = eval_expr(a, assignment) {
                changed |= propagate_expr(b, av == required, assignment);
            }
            if let Some(bv) = eval_expr(b, assignment) {
                changed |= propagate_expr(a, bv == required, assignment);
            }
            changed
        }
        Expr::Other(_) => false,
    }
}

/// Whether a `card` is mandatory (min >= 1): absent, `"+"`, or a range with min >= 1.
/// `"?"`/`"*"`/min-0 ranges are optional.
fn is_mandatory_card(card: &Option<String>) -> bool {
    match card.as_deref() {
        None => true,
        Some("+") => true,
        Some("?") | Some("*") => false,
        Some(s) => ncard_min(s).map(|min| min >= 1).unwrap_or(true),
    }
}

/// (min, max) bounds for a group cardinality. `max = None` means unbounded.
fn gcard_bounds(gcard: &str) -> (usize, Option<usize>) {
    match gcard {
        "xor" => (1, Some(1)),
        "mux" => (0, Some(1)),
        "or" => (1, None),
        "opt" => (0, None),
        s => {
            let min = ncard_min(s).unwrap_or(0);
            let max = ncard_max(s);
            (min, max)
        }
    }
}

/// Parses the min of a cardinality string like `"1..3"`, `"1..*"`, or `"2"`.
fn ncard_min(s: &str) -> Option<usize> {
    let min_part = s.split("..").next().unwrap_or(s);
    min_part.trim().parse::<usize>().ok()
}

/// Parses the max of a cardinality string: `"1..3"` -> `Some(3)`, `"1..*"` -> `None`,
/// bare `"2"` -> `Some(2)`.
fn ncard_max(s: &str) -> Option<usize> {
    match s.split_once("..") {
        Some((_, "*")) => None,
        Some((_, max)) => max.trim().parse::<usize>().ok(),
        None => s.trim().parse::<usize>().ok(),
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::parser::grammar::parse_clafer_module;

    const MODEL: &str = r#"
abstract POSSystem {
    xor Storage {
        SQLite
        Postgres
    }
    Scanner ?
    DigitalPayment ?
    PhysicalCash ?
    ReceiptPrinter ?

    // Dependencies
    [ PhysicalCash => ReceiptPrinter ]
    [ DigitalPayment || PhysicalCash ]
}

// Instance 1: A lightweight, digital-only tablet
MinimalistKiosk : POSSystem {
    [ SQLite ]
    [ DigitalPayment ]
    [ no PhysicalCash ]
    [ no Scanner ]
    [ no ReceiptPrinter ]
}

// Instance 2: A robust, hardware-heavy checkout
FullServiceStation : POSSystem {
    [ Postgres ]
    [ DigitalPayment ]
    [ PhysicalCash ]
    [ Scanner ]
}
"#;

    fn set(names: &[&str]) -> HashSet<String> {
        names.iter().map(|s| s.to_string()).collect()
    }

    #[test]
    fn resolves_full_service_station() {
        let decls = parse_clafer_module(MODEL);
        let eval = Evaluator::new(decls);
        let config = eval.resolve_instance("FullServiceStation").unwrap();
        assert_eq!(
            config.included,
            set(&["Postgres", "DigitalPayment", "PhysicalCash", "Scanner", "ReceiptPrinter"])
        );
        assert_eq!(config.excluded, set(&["SQLite"]));
    }

    #[test]
    fn resolves_minimalist_kiosk() {
        let decls = parse_clafer_module(MODEL);
        let eval = Evaluator::new(decls);
        let config = eval.resolve_instance("MinimalistKiosk").unwrap();
        assert_eq!(config.included, set(&["SQLite", "DigitalPayment"]));
        assert_eq!(
            config.excluded,
            set(&["Postgres", "PhysicalCash", "Scanner", "ReceiptPrinter"])
        );
    }

    #[test]
    fn unknown_instance_errors() {
        let decls = parse_clafer_module(MODEL);
        let eval = Evaluator::new(decls);
        assert!(eval.resolve_instance("NoSuchInstance").is_err());
    }
}

File: git_feature/domain/variability/evaluator.py
"""Three-valued instance resolution over the feature tree (port of the Rust evaluator)."""

from dataclasses import dataclass, field

from .ast import And, Clafer, Expr, Iff, Implies, Module, Not, Or, Ref, Xor, render

Assignment = dict[str, bool | None]


class ModelError(Exception):
    """The model itself is unusable for the request (unknown instance, missing super)."""


class ConfigurationError(Exception):
    """The requested configuration violates the model's constraints."""


@dataclass
class Configuration:
    """A resolved instance: concrete leaf features classified as included or excluded."""

    included: set[str] = field(default_factory=set)
    excluded: set[str] = field(default_factory=set)


class Evaluator:
    """Resolves named instances of an abstract feature model into configurations."""

    def __init__(self, module: Module) -> None:
        self.module = module
        self.registry: dict[str, Clafer] = {}
        self.duplicates: list[str] = []
        for clafer in module.clafers:
            self._index(clafer)

    def _index(self, clafer: Clafer) -> None:
        if clafer.name in self.registry:
            self.duplicates.append(clafer.name)
        self.registry[clafer.name] = clafer
        for child in clafer.children:
            self._index(child)

    def resolve_instance(self, instance_name: str) -> Configuration:
        """Classify the concrete leaf features of an instance as included/excluded."""
        assignment, supertype = self._resolve(instance_name)
        config = Configuration()
        for name in assignment:
            if name == supertype.name:
                continue
            clafer = self.registry.get(name)
            if clafer is None or clafer.is_abstract or clafer.children:
                continue
            if assignment.get(name) is True:
                config.included.add(name)
            else:
                config.excluded.add(name)
        return config

    def resolve_assignment(self, instance_name: str) -> dict[str, bool]:
        """Total feature → bool map over the model subtree (undetermined = excluded)."""
        assignment, supertype = self._resolve(instance_name)
        return {
            name: value is True for name, value in assignment.items() if name != supertype.name
        } | {supertype.name: True}

    def _resolve(self, instance_name: str) -> tuple[Assignment, Clafer]:
        instance = self.registry.get(instance_name)
        if instance is None:
            raise ModelError(f"instance '{instance_name}' not found in model")
        if instance.super_type is None:
            raise ModelError(f"instance '{instance_name}' has no super type")
        supertype = self.registry.get(instance.super_type)
        if supertype is None:
            raise ModelError(
                f"super type '{instance.super_type}' of instance '{instance_name}' not found"
            )

        subtree_names: set[str] = set()
        self._collect_subtree_names(supertype, subtree_names)
        assignment: Assignment = dict.fromkeys(subtree_names)
        conflicts: list[str] = []
        _try_assign(assignment, supertype.name, True, conflicts)

        # Explicit instance-body selections are seeded first so they win over inference.
        instance_constraints: list[Expr] = []
        _collect_constraints(instance, instance_constraints)
        for expr in instance_constraints:
            _propagate_expr(expr, True, assignment, conflicts)

        constraints = list(instance_constraints)
        _collect_constraints(supertype, constraints)

        while True:
            changed = self._propagate_mandatory_children(subtree_names, assignment, conflicts)
            changed |= self._propagate_group_cardinality(subtree_names, assignment, conflicts)
            for expr in constraints:
                changed |= _propagate_expr(expr, True, assignment, conflicts)
            if not changed:
                break

        self._check_legality(instance_name, constraints, assignment, conflicts, subtree_names)
        return assignment, supertype

    def _check_legality(
        self,
        instance_name: str,
        constraints: list[Expr],
        assignment: Assignment,
        conflicts: list[str],
        subtree_names: set[str],
    ) -> None:
        """Hard-fail on any violated constraint, conflicting inference, or broken group."""
        problems = list(conflicts)
        for expr in constraints:
            if _eval_expr(expr, assignment) is False:
                problems.append(f"constraint {render(expr)} is violated")
        total: Assignment = {name: value is True for name, value in assignment.items()}
        for name in subtree_names:
            clafer = self.registry.get(name)
            if clafer is None or clafer.gcard is None or total.get(name) is not True:
                continue
            active = sum(1 for child in clafer.children if total.get(child.name) is True)
            minimum, maximum = _gcard_bounds(clafer.gcard)
            if active < minimum or (maximum is not None and active > maximum):
                problems.append(
                    f"group '{name}' ({clafer.gcard}) has {active} active children,"
                    f" allowed {minimum}..{maximum if maximum is not None else '*'}"
                )
        if problems:
            details = "; ".join(problems)
            raise ConfigurationError(f"configuration '{instance_name}' is illegal: {details}")

    def _collect_subtree_names(self, root: Clafer, into: set[str]) -> None:
        into.add(root.name)
        for child in root.children:
            self._collect_subtree_names(child, into)

    def _propagate_mandatory_children(
        self, subtree_names: set[str], assignment: Assignment, conflicts: list[str]
    ) -> bool:
        changed = False
        for name in subtree_names:
            if assignment.get(name) is not True:
                continue
            clafer = self.registry.get(name)
            if clafer is None or clafer.gcard is not None:
                continue
            for child in clafer.children:
                if _is_mandatory_card(child.card):
                    changed |= _try_assign(assignment, child.name, True, conflicts)
        return changed

    def _propagate_group_cardinality(
        self, subtree_names: set[str], assignment: Assignment, conflicts: list[str]
    ) -> bool:
        changed = False
        for name in subtree_names:
            if assignment.get(name) is not True:
                continue
            clafer = self.registry.get(name)
            if clafer is None or clafer.gcard is None:
                continue
            minimum, maximum = _gcard_bounds(clafer.gcard)
            children = [child.name for child in clafer.children]
            true_count = sum(1 for child in children if assignment.get(child) is True)
            unknown = [child for child in children if assignment.get(child) is None]
            if maximum is not None and true_count >= maximum:
                for child in unknown:
                    changed |= _try_assign(assignment, child, False, conflicts)
                continue
            if minimum > 0 and true_count + len(unknown) == minimum:
                for child in unknown:
                    changed |= _try_assign(assignment, child, True, conflicts)
        return changed


def _collect_constraints(clafer: Clafer, out: list[Expr]) -> None:
    out.extend(clafer.constraints)
    for child in clafer.children:
        _collect_constraints(child, out)


def _try_assign(assignment: Assignment, name: str, value: bool, conflicts: list[str]) -> bool:
    """Assign only if currently unknown (monotonic); record conflicts and unknown names."""
    if name not in assignment:
        conflicts.append(f"reference to unknown feature '{name}'")
        return False
    current = assignment[name]
    if current is None:
        assignment[name] = value
        return True
    if current != value:
        conflicts.append(f"conflicting requirements for '{name}' (both on and off)")
    return False


def _eval_expr(expr: Expr, assignment: Assignment) -> bool | None:
    """Three-valued evaluation over a partial assignment."""
    match expr:
        case Ref(name):
            return assignment.get(name)
        case Not(operand):
            value = _eval_expr(operand, assignment)
            return None if value is None else not value
        case And(a, b):
            left, right = _eval_expr(a, assignment), _eval_expr(b, assignment)
            if left is False or right is False:
                return False
            if left is True and right is True:
                return True
            return None
        case Or(a, b):
            left, right = _eval_expr(a, assignment), _eval_expr(b, assignment)
            if left is True or right is True:
                return True
            if left is False and right is False:
                return False
            return None
        case Implies(a, b):
            left, right = _eval_expr(a, assignment), _eval_expr(b, assignment)
            if left is False or right is True:
                return True
            if left is True and right is False:
                return False
            return None
        case Xor(a, b):
            left, right = _eval_expr(a, assignment), _eval_expr(b, assignment)
            if left is None or right is None:
                return None
            return left != right
        case Iff(a, b):
            left, right = _eval_expr(a, assignment), _eval_expr(b, assignment)
            if left is None or right is None:
                return None
            return left == right


def _propagate_expr(
    expr: Expr, required: bool, assignment: Assignment, conflicts: list[str]
) -> bool:
    """Push a required truth value down the expression, asserting what follows."""
    match expr:
        case Ref(name):
            return _try_assign(assignment, name, required, conflicts)
        case Not(operand):
            return _propagate_expr(operand, not required, assignment, conflicts)
        case And(a, b):
            if required:
                changed = _propagate_expr(a, True, assignment, conflicts)
                return _propagate_expr(b, True, assignment, conflicts) or changed
            changed = False
            if _eval_expr(a, assignment) is True:
                changed |= _propagate_expr(b, False, assignment, conflicts)
            if _eval_expr(b, assignment) is True:
                changed |= _propagate_expr(a, False, assignment, conflicts)
            return changed
        case Or(a, b):
            if required:
                changed = False
                if _eval_expr(a, assignment) is False:
                    changed |= _propagate_expr(b, True, assignment, conflicts)
                if _eval_expr(b, assignment) is False:
                    changed |= _propagate_expr(a, True, assignment, conflicts)
                return changed
            changed = _propagate_expr(a, False, assignment, conflicts)
            return _propagate_expr(b, False, assignment, conflicts) or changed
        case Implies(a, b):
            if required:
                changed = False
                if _eval_expr(a, assignment) is True:
                    changed |= _propagate_expr(b, True, assignment, conflicts)
                if _eval_expr(b, assignment) is False:
                    changed |= _propagate_expr(a, False, assignment, conflicts)
                return changed
            changed = _propagate_expr(a, True, assignment, conflicts)
            return _propagate_expr(b, False, assignment, conflicts) or changed
        case Xor(a, b):
            changed = False
            left, right = _eval_expr(a, assignment), _eval_expr(b, assignment)
            if left is not None:
                changed |= _propagate_expr(b, left != required, assignment, conflicts)
            if right is not None:
                changed |= _propagate_expr(a, right != required, assignment, conflicts)
            return changed
        case Iff(a, b):
            changed = False
            left, right = _eval_expr(a, assignment), _eval_expr(b, assignment)
            if left is not None:
                changed |= _propagate_expr(b, left == required, assignment, conflicts)
            if right is not None:
                changed |= _propagate_expr(a, right == required, assignment, conflicts)
            return changed


def _is_mandatory_card(card: str | None) -> bool:
    """Mandatory means min >= 1: absent, '+', or a range/number with min >= 1."""
    if card is None or card == "+":
        return True
    if card in ("?", "*"):
        return False
    minimum = _ncard_min(card)
    return True if minimum is None else minimum >= 1


def _gcard_bounds(gcard: str) -> tuple[int, int | None]:
    """(min, max) for a group cardinality; max None means unbounded."""
    named = {"xor": (1, 1), "mux": (0, 1), "or": (1, None), "opt": (0, None)}
    if gcard in named:
        return named[gcard]
    return (_ncard_min(gcard) or 0, _ncard_max(gcard))


def _ncard_min(card: str) -> int | None:
    head = card.split("..")[0].strip()
    return int(head) if head.isdigit() else None


def _ncard_max(card: str) -> int | None:
    head, sep, tail = card.partition("..")
    part = tail.strip() if sep else head.strip()
    if part == "*":
        return None
    return int(part) if part.isdigit() else None
